src/adaptors/mod.rs

```diff
@@ -581,6 +581,17 @@ where
     fn size_hint(&self) -> (usize, Option<usize>) {
         (0, self.iter.size_hint().1)
     }
+
+    fn fold<B, F>(self, acc: B, f: F) -> B
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> B,
+    {
+        self.iter
+            .take_while(|opt| opt.is_some())
+            .map(|item| item.unwrap())
+            .fold(acc, f)
+    }
```

Comment:
I think it's possible to implement this without the `unwrap`, using `try_fold`:
```suggestion
    fn fold<B, F>(mut self, init: B, mut f: F) -> B
    where
        Self: Sized,
        F: FnMut(B, Self::Item) -> B,
    {
        // Replace with `ControlFlow`, once MSRV >= 1.55.0
        use core::result::Result::{Ok as Continue, Err as Break};
        let res = self.iter.try_fold(init, |acc, item| match item {
            Some(item) => Continue(f(acc, item)),
            None => Break(acc),
        });
        let (Break(res) | Continue(res)) = res;
        res
    }
```

Reply:
@jswrenn Thanks for your feedback. I'm a bit hesitate to override `fold` with `try_fold` though. 

The names of the methods in Rust are usually semantically clear. When overriding `fold`, it makes sense to delegate its responsibility to the underlying iterator's `fold`. Similarly, `try_fold` should be reserved for overriding `try_fold`. 

What do you think ?

Reply:
In the Rust standard library, it's quite common for `fold` to be implemented in terms of `try_fold`. In your original code you construct a `TakeWhile` and call `fold` on it. But `TakeWhile`'s fold implementation delegates to `try_fold`!

The general rule to follow is something like this:
- If you need to completely consume an iterator, use `for_each` or `fold`.
  - Use `for_each` if you don't need to maintain an accumulator.
  - Use `fold` if you do need to maintain an accumulator.
- If you need to partly consume an iterator, use `try_for_each` or `try_fold`.
  - Same rules about accumulators apply.

Here, `try_fold` is the perfect tool: We need to consume the iterator partially, up until the first `None`, with an accumulator.